From 56c068e22259361b8a2dcd89da5840cc0950b643 Mon Sep 17 00:00:00 2001
From: mintoyatsu <53593903+mintoyatsu@users.noreply.github.com>
Date: Tue, 10 Mar 2020 04:06:11 -0700
Subject: [PATCH] Server difficulty API

Allows plugins to get/set difficulty on a per-world basis.

diff --git a/src/main/java/org/bukkit/Difficulty.java b/src/main/java/org/bukkit/Difficulty.java
new file mode 100644
index 00000000..1e694f79
--- /dev/null
+++ b/src/main/java/org/bukkit/Difficulty.java
@@ -0,0 +1,61 @@
+package org.bukkit;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents the various difficulty levels that are available.
+ */
+public enum Difficulty {
+    /**
+     * Players regain health over time, hostile mobs don't spawn.
+     */
+    PEACEFUL(0),
+
+    /**
+     * Hostile mobs spawn, enemies deal less damage than on normal difficulty. (Default value)
+     */
+    EASY(1),
+
+    /**
+     * Hostile mobs spawn, enemies deal normal amounts of damage.
+     */
+    NORMAL(2),
+
+    /**
+     * Hostile mobs spawn, enemies deal greater damage than on normal difficulty.
+     */
+    HARD(3);
+
+    private final int value;
+    private final static Map<Integer, Difficulty> difficulties = new HashMap<Integer, Difficulty>();
+
+    private Difficulty(final int value) {
+        this.value = value;
+    }
+
+    /**
+     * Gets the difficulty value associated with this Difficulty.
+     *
+     * @return An integer value of this difficulty
+     */
+    public int getValue() {
+        return value;
+    }
+
+    /**
+     * Gets the Difficulty represented by the specified value
+     *
+     * @param value Value to check
+     * @return Associated {@link Difficulty} with the given value, or null if it doesn't exist
+     */
+    public static Difficulty getByValue(final int value) {
+        return difficulties.get(value);
+    }
+
+    static {
+        for (Difficulty diff : Difficulty.values()) {
+            difficulties.put(diff.getValue(), diff);
+        }
+    }
+}
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index c46e963a..c8f5ea5c 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -714,6 +714,22 @@ public interface World {
      */
     public void setAutoSave(boolean value);
 
+    // Canyon start - difficulty API
+    /**
+     * Sets the Difficulty of the world.
+     *
+     * @param difficulty the new difficulty you want to set the world to
+     */
+    public void setDifficulty(Difficulty difficulty);
+
+    /**
+     * Gets the Difficulty of the world.
+     *
+     * @return The difficulty of the world.
+     */
+    public Difficulty getDifficulty();
+    // Canyon end
+
     /**
      * Represents various map environment types that a world may be
      */
-- 
2.26.2

